pontoon/pretranslation/utils.py

```diff
@@ -0,0 +1,162 @@
+from six.moves import reduce
+import operator
+
+from django.utils import timezone
+from django.db.models import Q, CharField, Value as V
+from django.db.models.functions import Concat
+from bulk_update.helper import bulk_update
+
+import pontoon.base as base
+from pontoon.machinery.utils import get_google_translate_data, get_translation_memory_data
+
+
+def pretranslate(project, locales=None, entities=None):
+    """
+    Identifies strings without any translations and any suggestions.
+    Engages TheAlgorithm (bug 1552796) to gather pre-translations.
+    Stores pre-translations as suggestions (approved=False) to DB.
+    Author should be set to a name like "Pontoon Translator <pontoon@mozilla.com>".
+    Stats and Latest activity should be updated
+    """
+    if not locales:
+        locales = project.locales.filter(
+            project_locale__readonly=False,
+        ).prefetch_project_locale(project)
+
+    if not entities:
+        entities = base.models.Entity.objects.filter(
+            resource__project=project,
+            obsolete=False,
+        ).prefetch_related('resource')
+
+    # Fetch all distinct locale-entity pairs for which translation exists
+    translated_entities = base.models.Translation.objects.filter(
+        locale__in=locales,
+        entity__in=entities,
+    ).annotate(
+        locale_entity=Concat('locale_id', V('-'), 'entity_id', output_field=CharField())
+    ).values_list('locale_entity', flat=True).distinct()
+
+    translated_entities = list(translated_entities)
+
+    now = timezone.now()
+
+    user, _ = base.models.User.objects.get_or_create(
+        email="pontoon@mozilla.com",
+        username="pontoon-translator",
+        first_name="Pontoon Translator",
+    )
```

Comment:
We should create this user with a data migration. I believe however we should create separate users for MT and TM though (`first_name="Google Translate"` & `first_name="Translation Memory"`), because Google Translate API requires [attribution](https://cloud.google.com/translate/attribution#attribution_and_logos).